messageforums-app/src/java/org/sakaiproject/tool/messageforums/entityproviders/ForumsEntityProviderImpl.java

```diff
@@ -443,4 +439,64 @@ private void checkSiteAndToolAccess(String siteId) throws EntityException {
 			throw new EntityException("No access to tool in site: " + siteId, "",HttpServletResponse.SC_UNAUTHORIZED);
 		}
 	}
+
+	/**
+	 * Handles requests /direct/forums/messages/SITEID.json?n=10 and returns latest threads for a site
+	 * @param view
+	 * @param params
+	 * @return
+	 */
+	@EntityCustomAction(action="messages",viewKey=EntityView.VIEW_LIST)
+	public Object displayMessages(EntityView view, Map<String, Object> params) {
+
+		if(LOG.isDebugEnabled()) {
+			LOG.debug("handleSite");
+		}
+
+		//get number of messages to display from the URL params, validate and set to 0 if not set or conversion fails
+		int numberOfMessages = NumberUtils.toInt((String)params.get("n"), 0);
+		if(numberOfMessages == 0){
+			numberOfMessages = DEFAULT_NUM_MESSAGES;
+		}
+		String userId = developerHelperService.getCurrentUserId();
+
+		if(userId == null) {
+			LOG.error("Not logged in");
+			throw new EntityException("You must be logged in view conversations.","",HttpServletResponse.SC_UNAUTHORIZED);
+		}
+
+		String siteId = view.getPathSegment(2);
+
+		if(siteId == null) {
+			LOG.error("Bad request. No SITEID supplied on path.");
+			throw new EntityException("Bad request: To get the fora in a site you need a url like '/direct/forums/site/SITEID/latestmessages.json?n=10'"
+					,"",HttpServletResponse.SC_BAD_REQUEST);
+		}
+
+		checkSiteAndToolAccess(siteId);
+		List<SparseMessage> messages = new ArrayList<SparseMessage>();
+
+		for(Message fm : (List<Message>) forumManager.getRecentDiscussionForumMessagesByContext(siteId, numberOfMessages)) {
+			Topic topic = fm.getTopic();
+			//check read permission for the current user
+			if(!uiPermissionsManager.isRead(topic.getId(),((DiscussionTopic)topic).getDraft(),false, userId, siteId)) {
```

Comment:
But doesn't it have a count of how many messages are new? Or does it not actually check if the user can see those messages?